### ISSUE TITLE
How to write loss function with extra parameters?

### ISSUE THREAD

**Issue by jsh9** [DEAP/deap#331]
Can I write my own loss function that takes not only `individuals` but also `y_true` as parameters? In other words, I'd like my loss function to be something like this:

```{python}
def loss_func(individuals, y_true):
    y_predicted = (some manipulation of the parameters)
    loss = mean_absolute_error(y_predicted, y_true)
    return loss
```

Does DEAP currently support such type of loss functions? I couldn't find it in the documentations (or it is buried too deep it them). Thanks.
(My DEAP version is 1.2.)

**Comment by fmder** [comment]
How often does y_true change?

**Comment by jsh9** [comment]
It does not change during the "life time" of an optimization process, at least for my current case.
I'd like to avoid using global variables, or making the loss function a local function (because it's not picklable if a local function).

**Comment by fmder** [comment]
Then just register it on the toolbox as a keyword argument. The toolbox is
a wrapper on Python's functools.partial.

Cheers,

**Comment by chaltik** [comment]
what @fmder means is that if you don't need to change the values of `ytrue` at each step of evolution you just pass it in as a keyword arg while registering (and toolbox will use functools,partial to deal with it)
`
ytrue_predefined = from_somewhere
def loss_func(individuals,ytrue=None):
       loss=(some_manipulations)
       return loss

 toolbox.register("evaluate",
                     loss_func,
                     ytrue=ytrue_predefined
)`

**Comment by jsh9** [comment]
Thank you both @chaltik @fmder .
If I remember correctly, there is no example within DEAP that demonstrates this. Maybe you can consider adding this in a future release.

**Comment by mdalvi** [comment]
This is easy.

Here is my example on how I passed additional parameters to loss function,
```
def eval_fitness(individual, train_x, train_y, test_x, test_y):
   ......
```
And in toolbox you write
`toolbox.register('evaluate', eval_fitness, train_x=pca_X, train_y=y, test_x=pca_tX, test_y=ty)`

Done!

**Comment by Gabrielvon** [comment]
I did the following. Hope this helps.
```
def evaluate(individual, y_true):
    expr = str(individual)
    yhat = eval(expr, pset.context)
    if isinstance(yhat, (list, np.ndarray)):
        _fitness_value = scs.spearmanr(y_true, yhat).correlation
    else:
        _fitness_value = 0
    return _fitness_value,
y = (your true y values)
toolbox.register("evaluate", evaluate,  y_true=y)
```

**Comment by philip928lin** [comment]
I have a similar question but I need y_true to be different for each individual evaluation? Is it possible to do that? and how?
Thank you~~
